// @umijs/hooks/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useRequest", {
  enumerable: true,
  get: function get() {
    return _useRequest.default;
  }
});
Object.defineProperty(exports, "useAntdTable", {
  enumerable: true,
  get: function get() {
    return _useAntdTable.default;
  }
});
Object.defineProperty(exports, "useAPI", {
  enumerable: true,
  get: function get() {
    return _useAPI.default;
  }
});
Object.defineProperty(exports, "configRequest", {
  enumerable: true,
  get: function get() {
    return _useAPI.configRequest;
  }
});
Object.defineProperty(exports, "useAsync", {
  enumerable: true,
  get: function get() {
    return _useAsync.default;
  }
});
Object.defineProperty(exports, "useLoadMore", {
  enumerable: true,
  get: function get() {
    return _useLoadMore.default;
  }
});
Object.defineProperty(exports, "useSearch", {
  enumerable: true,
  get: function get() {
    return _useSearch.default;
  }
});
Object.defineProperty(exports, "useControllableValue", {
  enumerable: true,
  get: function get() {
    return _useControllableValue.default;
  }
});
Object.defineProperty(exports, "useDynamicList", {
  enumerable: true,
  get: function get() {
    return _useDynamicList.default;
  }
});
Object.defineProperty(exports, "useEventEmitter", {
  enumerable: true,
  get: function get() {
    return _useEventEmitter.default;
  }
});
Object.defineProperty(exports, "useVirtualList", {
  enumerable: true,
  get: function get() {
    return _useVirtualList.default;
  }
});
Object.defineProperty(exports, "configResponsive", {
  enumerable: true,
  get: function get() {
    return _useResponsive.configResponsive;
  }
});
Object.defineProperty(exports, "useResponsive", {
  enumerable: true,
  get: function get() {
    return _useResponsive.useResponsive;
  }
});
Object.defineProperty(exports, "useSize", {
  enumerable: true,
  get: function get() {
    return _useSize.default;
  }
});
Object.defineProperty(exports, "useLocalStorageState", {
  enumerable: true,
  get: function get() {
    return _useLocalStorageState.default;
  }
});
Object.defineProperty(exports, "useSessionStorageState", {
  enumerable: true,
  get: function get() {
    return _useSessionStorageState.default;
  }
});
Object.defineProperty(exports, "useUpdateEffect", {
  enumerable: true,
  get: function get() {
    return _useUpdateEffect.default;
  }
});
Object.defineProperty(exports, "useUpdateLayoutEffect", {
  enumerable: true,
  get: function get() {
    return _useUpdateLayoutEffect.default;
  }
});
Object.defineProperty(exports, "usePagination", {
  enumerable: true,
  get: function get() {
    return _usePagination.default;
  }
});
Object.defineProperty(exports, "useBoolean", {
  enumerable: true,
  get: function get() {
    return _useBoolean.default;
  }
});
Object.defineProperty(exports, "useToggle", {
  enumerable: true,
  get: function get() {
    return _useToggle.default;
  }
});
Object.defineProperty(exports, "useDocumentVisibility", {
  enumerable: true,
  get: function get() {
    return _useDocumentVisibility.default;
  }
});
Object.defineProperty(exports, "useSelections", {
  enumerable: true,
  get: function get() {
    return _useSelections.default;
  }
});
Object.defineProperty(exports, "useThrottle", {
  enumerable: true,
  get: function get() {
    return _useThrottle.default;
  }
});
Object.defineProperty(exports, "useThrottleFn", {
  enumerable: true,
  get: function get() {
    return _useThrottleFn.default;
  }
});
Object.defineProperty(exports, "useDebounce", {
  enumerable: true,
  get: function get() {
    return _useDebounce.default;
  }
});
Object.defineProperty(exports, "useDebounceFn", {
  enumerable: true,
  get: function get() {
    return _useDebounceFn.default;
  }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function get() {
    return _usePrevious.default;
  }
});
Object.defineProperty(exports, "useMouse", {
  enumerable: true,
  get: function get() {
    return _useMouse.default;
  }
});
Object.defineProperty(exports, "useScroll", {
  enumerable: true,
  get: function get() {
    return _useScroll.default;
  }
});
Object.defineProperty(exports, "useClickAway", {
  enumerable: true,
  get: function get() {
    return _useClickAway.default;
  }
});
Object.defineProperty(exports, "useFullscreen", {
  enumerable: true,
  get: function get() {
    return _useFullscreen.default;
  }
});
Object.defineProperty(exports, "useInViewport", {
  enumerable: true,
  get: function get() {
    return _useInViewport.default;
  }
});
Object.defineProperty(exports, "useKeyPress", {
  enumerable: true,
  get: function get() {
    return _useKeyPress.default;
  }
});
Object.defineProperty(exports, "useEventListener", {
  enumerable: true,
  get: function get() {
    return _useEventListener.default;
  }
});
Object.defineProperty(exports, "useHover", {
  enumerable: true,
  get: function get() {
    return _useHover.default;
  }
});
Object.defineProperty(exports, "useUnmount", {
  enumerable: true,
  get: function get() {
    return _useUnmount.default;
  }
});
Object.defineProperty(exports, "useFormTable", {
  enumerable: true,
  get: function get() {
    return _useFormTable.default;
  }
});
Object.defineProperty(exports, "useSet", {
  enumerable: true,
  get: function get() {
    return _useSet.default;
  }
});
Object.defineProperty(exports, "usePersistFn", {
  enumerable: true,
  get: function get() {
    return _usePersistFn.default;
  }
});
Object.defineProperty(exports, "useMap", {
  enumerable: true,
  get: function get() {
    return _useMap.default;
  }
});
Object.defineProperty(exports, "useCreation", {
  enumerable: true,
  get: function get() {
    return _useCreation.default;
  }
});
Object.defineProperty(exports, "useDrag", {
  enumerable: true,
  get: function get() {
    return _useDrop.useDrag;
  }
});
Object.defineProperty(exports, "useDrop", {
  enumerable: true,
  get: function get() {
    return _useDrop.useDrop;
  }
});
Object.defineProperty(exports, "useMount", {
  enumerable: true,
  get: function get() {
    return _useMount.default;
  }
});
Object.defineProperty(exports, "useTextSelection", {
  enumerable: true,
  get: function get() {
    return _useTextSelection.default;
  }
});
Object.defineProperty(exports, "useCounter", {
  enumerable: true,
  get: function get() {
    return _useCounter.default;
  }
});
Object.defineProperty(exports, "useUpdate", {
  enumerable: true,
  get: function get() {
    return _useUpdate.default;
  }
});
Object.defineProperty(exports, "useEventTarget", {
  enumerable: true,
  get: function get() {
    return _useEventTarget.default;
  }
});
Object.defineProperty(exports, "useHistoryTravel", {
  enumerable: true,
  get: function get() {
    return _useHistoryTravel.default;
  }
});
exports.useControlledValue = void 0;

var _useRequest = _interopRequireDefault(require("@umijs/use-request"));

var _useAntdTable = _interopRequireDefault(require("./useAntdTable"));

var _useAPI = _interopRequireWildcard(require("./useAPI"));

var _useAsync = _interopRequireDefault(require("./useAsync"));

var _useLoadMore = _interopRequireDefault(require("./useLoadMore"));

var _useSearch = _interopRequireDefault(require("./useSearch"));

var _useControllableValue = _interopRequireDefault(require("./useControllableValue"));

var _useDynamicList = _interopRequireDefault(require("./useDynamicList"));

var _useEventEmitter = _interopRequireDefault(require("./useEventEmitter"));

var _useVirtualList = _interopRequireDefault(require("./useVirtualList"));

var _useResponsive = require("./useResponsive");

var _useSize = _interopRequireDefault(require("./useSize"));

var _useLocalStorageState = _interopRequireDefault(require("./useLocalStorageState"));

var _useSessionStorageState = _interopRequireDefault(require("./useSessionStorageState"));

var _useUpdateEffect = _interopRequireDefault(require("./useUpdateEffect"));

var _useUpdateLayoutEffect = _interopRequireDefault(require("./useUpdateLayoutEffect"));

var _usePagination = _interopRequireDefault(require("./usePagination"));

var _useBoolean = _interopRequireDefault(require("./useBoolean"));

var _useToggle = _interopRequireDefault(require("./useToggle"));

var _useDocumentVisibility = _interopRequireDefault(require("./useDocumentVisibility"));

var _useSelections = _interopRequireDefault(require("./useSelections"));

var _useThrottle = _interopRequireDefault(require("./useThrottle"));

var _useThrottleFn = _interopRequireDefault(require("./useThrottleFn"));

var _useDebounce = _interopRequireDefault(require("./useDebounce"));

var _useDebounceFn = _interopRequireDefault(require("./useDebounceFn"));

var _usePrevious = _interopRequireDefault(require("./usePrevious"));

var _useMouse = _interopRequireDefault(require("./useMouse"));

var _useScroll = _interopRequireDefault(require("./useScroll"));

var _useClickAway = _interopRequireDefault(require("./useClickAway"));

var _useFullscreen = _interopRequireDefault(require("./useFullscreen"));

var _useInViewport = _interopRequireDefault(require("./useInViewport"));

var _useKeyPress = _interopRequireDefault(require("./useKeyPress"));

var _useEventListener = _interopRequireDefault(require("./useEventListener"));

var _useHover = _interopRequireDefault(require("./useHover"));

var _useUnmount = _interopRequireDefault(require("./useUnmount"));

var _useFormTable = _interopRequireDefault(require("./useFormTable"));

var _useSet = _interopRequireDefault(require("./useSet"));

var _usePersistFn = _interopRequireDefault(require("./usePersistFn"));

var _useMap = _interopRequireDefault(require("./useMap"));

var _useCreation = _interopRequireDefault(require("./useCreation"));

var _useDrop = require("./useDrop");

var _useMount = _interopRequireDefault(require("./useMount"));

var _useTextSelection = _interopRequireDefault(require("./useTextSelection"));

var _useCounter = _interopRequireDefault(require("./useCounter"));

var _useUpdate = _interopRequireDefault(require("./useUpdate"));

var _useEventTarget = _interopRequireDefault(require("./useEventTarget"));

var _useHistoryTravel = _interopRequireDefault(require("./useHistoryTravel"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var useControlledValue = function useControlledValue() {
  console.warn('useControlledValue is deprecated and will be removed in the next major version. Please use useControllableValue instead.');
  return _useControllableValue.default.apply(void 0, arguments);
};

exports.useControlledValue = useControlledValue;
