// @umijs/hooks/es/useEventListener/index.js
import { useEffect, useRef } from 'react';

function useEventListener(eventName, handler, options) {
  var ref = useRef();
  var savedHandler = useRef();
  useEffect(function () {
    savedHandler.current = handler;
  }, [handler]);
  useEffect(function () {
    var passedInElement = options && (typeof options.dom === 'function' ? options.dom() : options.dom);
    var element = passedInElement ? passedInElement : ref.current || window;
    var isSupported = element.addEventListener;
    if (!isSupported) return;

    var eventListener = function eventListener(event) {
      return savedHandler.current && savedHandler.current(event);
    };

    element.addEventListener(eventName, eventListener, {
      capture: options === null || options === void 0 ? void 0 : options.capture,
      once: options === null || options === void 0 ? void 0 : options.once,
      passive: options === null || options === void 0 ? void 0 : options.passive
    });
    return function () {
      element.removeEventListener(eventName, eventListener, {
        capture: options === null || options === void 0 ? void 0 : options.capture
      });
    };
  }, [eventName, options, ref.current]);
  return ref;
}

export default useEventListener;
