// @umijs/hooks/lib/useEventEmitter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useEventEmitter;
exports.EventEmitter = void 0;

var _react = require("react");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var EventEmitter = function EventEmitter() {
  var _this = this;

  _classCallCheck(this, EventEmitter);

  this.subscriptions = new Set();

  this.emit = function (val) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _this.subscriptions[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var subscription = _step.value;
        subscription(val);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  };

  this.useSubscription = function (callback) {
    var callbackRef = (0, _react.useRef)();
    callbackRef.current = callback;
    (0, _react.useEffect)(function () {
      function subscription(val) {
        if (callbackRef.current) {
          callbackRef.current(val);
        }
      }

      _this.subscriptions.add(subscription);

      return function () {
        _this.subscriptions.delete(subscription);
      };
    }, []);
  };
};

exports.EventEmitter = EventEmitter;

function useEventEmitter() {
  var ref = (0, _react.useRef)();

  if (!ref.current) {
    ref.current = new EventEmitter();
  }

  return ref.current;
}
